src/Passes/Source/ValidationPass/ValidationPass.cpp

```diff
@@ -0,0 +1,115 @@
+                    if (opcodes_.find(opname) != opcodes_.end())
+                    {
+                        ++opcodes_[opname];
+                    }
+                    else
+                    {
+                        opcodes_[opname] = 1;
+                    }
+                    auto call_instr = llvm::dyn_cast<llvm::CallBase>(&instr);
+                    if (call_instr != nullptr)
+                    {
+                        auto f = call_instr->getCalledFunction();
+                        if (f == nullptr)
+                        {
+                            continue;
+                        }
+
+                        auto name = static_cast<std::string>(f->getName());
+                        if (f->isDeclaration())
+                        {
+                            if (external_calls_.find(name) != external_calls_.end())
+                            {
+                                ++external_calls_[name];
+                            }
+                            else
+                            {
+                                external_calls_[name] = 1;
+                            }
+                        }
+                        else
+                        {
+                            if (internal_calls_.find(name) != internal_calls_.end())
+                            {
+                                ++internal_calls_[name];
+                            }
+                            else
+                            {
+                                internal_calls_[name] = 1;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        bool raise_exception = false;
+        if (config_.whitelistOpcodes())
+        {
+            auto const& allowed_ops = config_.allowedOpcodes();
+            for (auto const& k : opcodes_)
+            {
+                if (allowed_ops.find(k.first) == allowed_ops.end())
+                {
+                    llvm::errs() << "; Error: '" << k.first << "' is not allowed for this profile.\n";
+                }
+            }
+        }
+
+        auto const& allowed_functions = config_.allowedExternalCallNames();
+        if (config_.whitelistOpcodes())
+        {
+            for (auto const& k : external_calls_)
+            {
+                if (allowed_functions.find(k.first) == allowed_functions.end())
+                {
+                    llvm::errs() << "; Error: '" << k.first << "' is not allowed for this profile.\n";
+                }
+            }
+        }
+
+        if (!config_.allowInternalCalls() && !internal_calls_.empty())
+        {
+            llvm::errs() << "; Error: Calls to custom defined functions not allowed\n";
+            raise_exception = true;
+        }
+
+        if (raise_exception)
+        {
+            throw std::runtime_error("QIR is not valid within the defined profile");
+        }
```

Comment:
I fail to see where this function (`ValidationPass::run()`) is called. 
I assume 
* it is called by the LLVM mechanisms;
* it is the LLVM mechanisms who expect this function to throw an exception in case of validation failure, 
and the LLVM mechanisms handle the exception appropriately.

Otherwise the situation seems to me in risk of a crash (upon unhandled exception) in case of validation failure.